[ONNX] Wrap op validation inputs and add export_options.py and function_dispatcher.py

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94721
* #94720

1. `_validate_op_between_ort_torch` inputs was not wrapped (preprocessed) properly.
2. Introduce function_dispatcher.py to store decompistion table (atn/prim) and ATenLib
3. Introduce ~~export_options.py~~ options.py

titaiwangms committed:
[ONNX] Wrap op validation inputs and add fx_utils.py

[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/fx/fx_utils.py:
```diff
@@ -0,0 +1,287 @@
+"""Utilities for manipulating the torch.fx object."""
```
I am inclined to find a more descriptive name than utils. Because once something is called utils, everything pours in. In particular, I wouldn't consider the `_ATENLIB_FUNCTIONS` a utility for onnx fx.

BowenBao replied:
Looks like the file contains the following components that can be further divided.
* ops registration and dispatching.
* decomposition.
* export option.

titaiwangms replied:
Would adding function_dispatcher.py for atenlib and decomposition tables and export_options.py for ExportOptions be satisfactory? And move op validation back to exporter.py

justinchuby commented on torch/onnx/_internal/fx/fx_utils.py:
```diff
@@ -0,0 +1,287 @@
+"""Utilities for manipulating the torch.fx object."""
```
Also: is `Utilities for manipulating the torch.fx object.` an accurate summary?

titaiwangms committed:
Update on "[ONNX] Wrap op validation inputs and add fx_utils.py"


`_validate_op_between_ort_torch` inputs was not wrapped (preprocessed) properly

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/export_options.py:
```diff
@@ -0,0 +1,33 @@
+"""Options for FX exporter static kwargs."""
+from __future__ import annotations
+
```
"export_options.py" -> "options.py"?

titaiwangms replied:
Tried to move verification options inthis too, but failed due to https://github.com/pytorch/pytorch/issues/94800

BowenBao replied:
Actually ... I meant file renaming lol. Sry

titaiwangms replied:
I am moving VerificationOptions also into options.py (and move whole thing outside of fx, but under _internal.), and it turns out the name options.py is too general, as most of our variables/parameters are already named options. So I will just go with kwarg_options,py. What do you think?

BowenBao commented on torch/onnx/_internal/fx/function_dispatcher.py:
```diff
@@ -0,0 +1,205 @@
+"""Dispatcher for torch ATen decomposition and AtenLib functions from onnx-script."""
```
```suggestion
"""Dispatcher for AtenLib functions from onnx-script."""
```
Not in this PR, decomposition table stuff should eventually be moved to fx frontend side.

titaiwangms replied:
https://github.com/pytorch/pytorch/issues/94800

BowenBao commented on torch/onnx/_internal/fx/export_options.py:
```diff
@@ -0,0 +1,33 @@
+"""Options for FX exporter static kwargs."""
```
```suggestion
"""Options for FX exporter."""
```
I wonder what 'static kwargs' implies/adds...

titaiwangms committed:
Update on "[ONNX] Wrap op validation inputs and add export_options.py and function_dispatcher.py"


1. `_validate_op_between_ort_torch` inputs was not wrapped (preprocessed) properly.
2. Introduce function_dispatcher.py to store decompistion table (atn/prim) and ATenLib
3. Introduce export_options.py 

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Wrap op validation inputs and add export_options.py and function_dispatcher.py"


1. `_validate_op_between_ort_torch` inputs was not wrapped (preprocessed) properly.
2. Introduce function_dispatcher.py to store decompistion table (atn/prim) and ATenLib
3. Introduce export_options.py 

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge -g

titaiwangms committed:
Update on "[ONNX] Wrap op validation inputs and add export_options.py and function_dispatcher.py"


1. `_validate_op_between_ort_torch` inputs was not wrapped (preprocessed) properly.
2. Introduce function_dispatcher.py to store decompistion table (atn/prim) and ATenLib
3. Introduce ~~export_options.py~~ options.py

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge -g
